src/ace/managers/state.c

```diff
@@ -0,0 +1,136 @@
+	logBlockBegin("stateManagerDestroy(pStateManager: %p)", pStateManager);
+
+	statePopAll(pStateManager);
+
+	memFree(pStateManager, sizeof(tStateManager));
+
+	logBlockEnd("stateManagerDestroy()");
+}
+
+tState *stateCreate(
+	tStateCb cbCreate, tStateCb cbLoop, tStateCb cbDestroy,
+	tStateCb cbSuspend, tStateCb cbResume,
+	tState *pPrev
+) {
+	logBlockBegin(
+		"stateCreate(cbCreate: %p, cbLoop: %p, cbDestroy: %p, cbSuspend: %p, cbResume: %p, pPrev: %p)",
+		cbCreate, cbLoop, cbDestroy, cbSuspend, cbResume, pPrev
+	);
+
+	tState *pState = memAllocFast(sizeof(tState));
+
+	pState->cbCreate = cbCreate;
+	pState->cbLoop = cbLoop;
+	pState->cbDestroy = cbDestroy;
+	pState->cbSuspend = cbSuspend;
+	pState->cbResume = cbResume;
+	pState->pPrev = pPrev;
+
+	logBlockEnd("stateCreate()");
+
+	return pState;
+}
+
+void stateDestroy(tState *pState) {
+	logBlockBegin("stateDestroy(pState: %p)", pState);
+	
+	memFree(pState, sizeof(tState));
+
+	logBlockEnd("stateDestroy()");
+}
+
+void statePush(tStateManager *pStateManager, tState *pState) {
+	logBlockBegin(
+		"statePush(pStateManager: %p, pState: %p)",
+		pStateManager, pState
+	);
+
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbSuspend) {
+		pStateManager->pCurrent->cbSuspend();
+	}
+
+	pState->pPrev = pStateManager->pCurrent;
+	pStateManager->pCurrent = pState;
+
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbCreate) {
+		pStateManager->pCurrent->cbCreate();
+	}
+
+	logBlockEnd("statePush()");
+}
+
+void statePop(tStateManager *pStateManager) {
+	logBlockBegin("statePop(pStateManager: %p)", pStateManager);
+
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbDestroy) {
+		pStateManager->pCurrent->cbDestroy();
+	}
+
+	tState *pOldState = pStateManager->pCurrent;
+	pStateManager->pCurrent = pOldState->pPrev;
+
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbResume) {
+		pStateManager->pCurrent->cbResume();
+	}
+
+	logBlockEnd("statePop()");
+}
+
+void statePopAll(tStateManager *pStateManager) {
+	logBlockBegin("statePopAll(pStateManager: %p)", pStateManager);
+
+	while (pStateManager->pCurrent) {
+		if (pStateManager->pCurrent->cbDestroy) {
+			pStateManager->pCurrent->cbDestroy();
+		}
+
+		pStateManager->pCurrent = pStateManager->pCurrent->pPrev;
+	}
+
+	logBlockEnd("statePopAll()");
+}
+
+void stateChange(tStateManager *pStateManager, tState *pState) {
+	logBlockBegin(
+		"stateChange(pStateManager: %p, pState: %p)",
+		pStateManager, pState
+	);
+
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbDestroy) {
+		pStateManager->pCurrent->cbDestroy();
+	}
+
+	pStateManager->pCurrent = pState;
+
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbCreate) {
+		pStateManager->pCurrent->cbCreate();
+	}
+
+	logBlockEnd("stateChange()");
+}
+
+void stateProcess(tStateManager *pStateManager) {
+	if (pStateManager->pCurrent && pStateManager->pCurrent->cbLoop) {
+		pStateManager->pCurrent->cbLoop();
+	}
+}
```

Comment:
configure vscode to put newline at the end of file. It's common practice in C world, it may produce compiler warnings in other version. Also, IIRC vbcc had problem with lack of newlines. Not that we support it, but it's a good practice. Even GH shows fancy icon that you've ommited it. Almost all of POSIX world uses such convention. https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file